lib/backend/k8s/syncer.go

```diff
@@ -0,0 +1,536 @@
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	k8sapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	"k8s.io/kubernetes/pkg/watch"
+	"reflect"
+)
+
+func newSyncer(kc KubeClient, callbacks api.SyncerCallbacks) *kubeSyncer {
+	return &kubeSyncer{
+		kc:        kc,
+		callbacks: callbacks,
+	}
+}
+
+type kubeSyncer struct {
+	kc        KubeClient
+	callbacks api.SyncerCallbacks
+	OneShot   bool
+}
+
+// Holds resource version information.
+type resourceVersions struct {
+	podVersion           string
+	namespaceVersion     string
+	networkPolicyVersion string
+}
+
+func getUpdateType(e watch.Event) model.UpdateType {
+	// Determine the update type.
+	updateType := model.UpdateTypeKVUnknown
+	switch e.Type {
+	case watch.Added:
+		updateType = model.UpdateTypeKVNew
+	case watch.Deleted:
+		updateType = model.UpdateTypeKVDeleted
+	case watch.Modified:
+		updateType = model.UpdateTypeKVUpdated
+	}
+	return updateType
+}
+
+func (syn *kubeSyncer) Start() {
+	// Channel for receiving updates from a snapshot.
+	snapshotUpdates := make(chan *[]model.Update)
+
+	// Channel for receiving updates from the watcher.
+	watchUpdates := make(chan *model.Update)
+
+	// Channel used by the watcher to trigger a re-sync.
+	triggerResync := make(chan *resourceVersions, 5)
+
+	// Channel to send the index from which to start the snapshot.
+	initialSnapshotIndex := make(chan *resourceVersions)
+
+	// Status channel
+	statusUpdates := make(chan api.SyncStatus)
+
+	// If we're not in one-shot mode, start the API watcher to
+	// gather updates.
+	if !syn.OneShot {
+		go syn.watchKubeAPI(watchUpdates, triggerResync, initialSnapshotIndex)
+	}
+
+	// Start a background thread to read snapshots from etcd.  It will
+	// read a start-of-day snapshot and then wait to be signalled on the
+	// resyncIndex channel.
+	go syn.readSnapshot(snapshotUpdates, triggerResync, initialSnapshotIndex, statusUpdates)
+
+	// Start a routine to merge updates from the snapshot routine and the
+	// watch routine (if running), and pass information to callbacks.
+	// TODO: We're never actually running both the snapshot thread and the watch thread at the same time
+	// so what's the point?
+	go syn.mergeUpdates(snapshotUpdates, watchUpdates, statusUpdates)
+}
+
+// mergeUpdates ensures that callbacks are all executed from the same thread.
+func (syn *kubeSyncer) mergeUpdates(snapshotUpdates chan *[]model.Update, watchUpdates chan *model.Update, statusUpdates chan api.SyncStatus) {
+	var update *model.Update
+	var updates *[]model.Update
+	currentStatus := api.WaitForDatastore
+	newStatus := api.WaitForDatastore
+	syn.callbacks.OnStatusUpdated(api.WaitForDatastore)
+	for {
+		select {
+		case updates = <-snapshotUpdates:
+			log.Debugf("Snapshot update: %+v", updates)
+			if currentStatus != api.ResyncInProgress {
+				panic("Recieved snapshot update while not resyncing")
+			}
+			syn.callbacks.OnUpdates(*updates)
```

Comment:
@fasaxc does Felix treat a snapshot different than any other batch of updates it might receive?

I would have thought that Felix should treat a snapshot as a "this is the current state", wiping out any any state it's storing internally and replacing it with the snapshot (so if an update isn't in the snapshot, but was known by Felix before the snapshot, it is essentially deleted from Felix upon receiving the snapshot)

For the Kubernetes case, if a resource is deleted _during_ a snapshot, the Kubernetes API will inform us of any deletes that occurred when we start the watches below, since we start with the provided `resourceVersion` from the snapshot for each resource type.

From Felix's point of view this means it will get a snapshot with a bunch of stuff in it, and then subsequent modify / delete events immediately following.

If the snapshot takes so long that the resourceVersions are immediately stale, then a new snapshot will be triggered right away.

Am I missing something?